### DISCUSSION TITLE
A problem, apparently, in the EXTEND instruction

### DISCUSSION THREAD

**Discussion by yishaybg** [virtualagc/virtualagc#1210] [Q&A] [answered]
I looked at the original AGC blueprints, and I noticed a problem, apparently, in the EXTEND instruction.

For example, if my code is:
....
....
EXTEND
SU
....
....
So at the subinstruction before the EXTEND (could be STD2, TC0, TCF0, RSM3, MP3, BZF0, or BZMF0), at time pulse T07 PHS4, the signal EXTPLS is produced, and flip-flop FUTEXT (FF30109-30110 in module A3) is set.
At time pulse T08 of these fetching subinstructions, control pulse RAD is produced and converted into signal RADRZ. Signal RADRZ is then converted into control pulses RZ and ST2 which cause subinstruction STD2 to be executed.
At time pulse T12, SQEXT flip-flop is set because NISQL is produced.

The next subinstruction is STD2.

At time pulse T01 of STD2, FUTEXT is reset, because INKBT1 is produced. At time pulse T12, SQEXT is reset too because in STD2, NISQL is set from T02 to T12.

The problem is that now, at the next subinstruction, that sould be an extracode instruction, the SQEXT will be reset, so it won't be an extracode instruction.

Am I correct? Am I missing something?

**Comment by thewonderidiot** [comment, marked as answer]
Hey, that's a good question! What you've missed, I think, is an expander gate hiding on [sheet 2 of the module A21 schematics](http://klabs.org/history/ech/agc_schematics/logic/a21-2.jpg).
![image](https://github.com/virtualagc/virtualagc/assets/94056/4042158a-86dd-4385-a500-b3866f80c272)

Its presence there suggests that this might have been an oversight in the original logic design that needed to be patched using a spare gate. This gate inhibits INKBT1 during STD2 subinstructions, which allows FUTEXT to persist throughout STD2. Here's the timing of the signals in question in our hardware simulator:
![image](https://github.com/virtualagc/virtualagc/assets/94056/62f850a2-9171-4e19-b069-fd87ebed12d5)

**Comment by thewonderidiot** [comment]
It's hard to read the dates, but it appears that this gate was a spare gate all the way up through revision D of the 2003100 variant of the logic schematics. It wasn't allocated for this purpose until revision E, sometime in late-ish 1965.

[Revision D:](https://archive.org/details/apertureCardBox459NARASW_images/apertureCardBox459NARASW/page/n1709/mode/1up?view=theater)
![image](https://github.com/virtualagc/virtualagc/assets/94056/128f304c-7ef2-4771-af59-5b9f6b23df2d)

[Revision E:](https://archive.org/details/apertureCardBox459NARASW_images/apertureCardBox459NARASW/page/n1712/mode/1up?view=theater)
![image](https://github.com/virtualagc/virtualagc/assets/94056/669bf8aa-ae5b-4435-aef4-00dbc6f70c1c)

**Reply by yishaybg** [reply]
Thank you very much!
How did extracode work until revision E?

**Reply by thewonderidiot** [reply]
I think the most likely answer is it didn't. :slightly_smiling_face: They got the first Block II computer (AGC 200M) working in December 1965, so while this was a late logic design change it likely happened before the bug made it to actual hardware.

**Reply by yishaybg** [reply]
Thanks!
